src/main/java/com/rovio/ingest/util/MetadataUpdater.java

```diff
@@ -91,50 +89,61 @@ private void testDbConnection() {
-            }
-
-            int[] execute = preparedBatch.execute();
-            if (execute.length != dataSegments.size()) {
-                throw new IllegalStateException(format("Failed to update All segments, segment=%d, updated=%d",
-                        dataSegments.size(), execute.length));
-            }
+        try {
+            RetryUtils.Task<Void> task = () -> {
+                metadataSegmentUpdaterJobHandler.publishSegments(segmentsTable, dataSegments, MAPPER);
+                return null;
+            };
+
+            RetryUtils.retry(
+                    task,
+                    this::isLockTimeoutException,
+                    0,
+                    DEFAULT_MAX_RETRIES,
+                    null,
+                    String.format("Lock wait timeout exception while publishing segments for %s", dataSource));
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to publish segments with retry", e);
+        }
 
-            Update updateStatement;
-            if (initDataSource) {
-                updateStatement = handle.createStatement(format(MARK_ALL_OLDER_SEGMENTS_AS_UNUSED_SQL, this.segmentsTable))
-                        .bind("dataSource", this.dataSource)
-                        .bind("version", this.version);
-                updateStatement.execute();
+        if (initDataSource) {
+            // Mark older version segments as unused as an additional step.
+            // Druid Coordinator already does this on every refreshInterval but this is done here just to speed up things.
+            try {
+                RetryUtils.Task<Integer> task = () -> this.sqlConnector.getDBI().withHandle(handle -> {
+                    Update updateStatement = handle.createStatement(format(MARK_ALL_OLDER_SEGMENTS_AS_UNUSED_SQL, this.segmentsTable))
+                            .bind("dataSource", this.dataSource)
+                            .bind("version", this.version);
+                    return updateStatement.execute();
+                });
+
+                RetryUtils.retry(
+                        task,
+                        this::isLockTimeoutException,
+                        0,
+                        DEFAULT_MAX_RETRIES,
+                        null,
+                        String.format("Lock wait timeout exception while marking old segments as unused with init for %s", dataSource));
+            } catch (Exception e) {
+                LOG.warn("Failed to mark old segments as unused with init after retries; ignored", e);
             }
-
-            handle.commit();
-            return execute;
-        });
+        }
+      }
+
+    private boolean isLockTimeoutException(Throwable e) {
+        return e != null &&
+                ((e instanceof BatchUpdateException &&
+                        e.getMessage() != null && e.getMessage().contains("Lock wait timeout exceeded")) ||
+                        isLockTimeoutException(e.getCause())
+                );
```

Comment:
Have you tested this condition to successfully "kick in"? For example, use some very short timeout temporarily to simulate a failure scenario.